File: Data Structures & Algorithms/balanced-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isBalanced(TreeNode* root) {
        
        bool ans = true;
        auto dfs = [&](auto& dfs, TreeNode* node) -> int {
            if (!node) return 0;

            if (!node->left && !node->right) return 0;

            int left = 0, right = 0;
            if (node->left) {
                left = 1 + dfs(dfs, node->left);
            }
            if (node->right) {
                right = 1 + dfs(dfs, node->right);
            }

            if (abs(left - right) > 1) {
                ans = false;
            }
            return max(left, right);
        };

        dfs(dfs, root);

        return ans;
    }
};

File: Data Structures & Algorithms/count-number-of-islands/submission-0.java
class Solution {
    int[][] dirs = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
    int n, m;
    public int numIslands(char[][] grid) {
        n = grid.length;
        m = grid[0].length;

        int ans = 0;
        boolean[][] vis = new boolean[n][m];
        for (int i=0; i<n; i++) {
            for (int j=0; j<m; j++) {
                if (!vis[i][j] && grid[i][j] == '1') {
                    ans++;
                    dfs(grid, i, j, vis);
                }
            }
        }
        return ans;
    }

    void dfs(char[][] grid, int i, int j, boolean[][] vis) {
        vis[i][j] = true;
        for (int[] d: dirs) {
            int di = d[0], dj = d[1];
            int ni = i+di, nj = j+dj;
            if (ni >= 0 && ni < n && nj >= 0 && nj < m && !vis[ni][nj] && grid[ni][nj] == '1') {
                dfs(grid, ni, nj, vis);
            }
        }        
    }
}
